ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -0,0 +1,201 @@
+function admin_ack() {
+    set -x
+    if [[ $1 -eq $2 ]]; then
+        echo "Upgrade between z-stream version does not require admin ack"
+    else
+        local out; out="$(./oc -n openshift-config-managed get configmap admin-gates -o json | jq -r ".data")"
+        if [[ $out == *"ack-4.$1"* ]]; then
+            echo "Require admin ack"
+            local ack_data; ack_data="$(echo "$out" | cut -f2 -d\")" && echo "Admin ack patch data is: ${ack_data}"
+            if ! ./oc -n openshift-config patch configmap admin-acks --patch '{"data":{"'"${ack_data}"'": "true"}}' --type=merge; then
+                echo >&2 "Admin-acks patch command failed, exiting" && exit 1
+            fi
+            
+            echo -e "sleep 5 min wait admin-acks patch to be valid...\n"
+            local wait_time_loop_var=0
+            while(( wait_time_loop_var < 5 ));do
+                sleep 1m
+                echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+                if ! ./oc adm upgrade|grep "AdminAckRequired"; then
+                    echo -e "Admin-acks patch still in processing, waiting...\n"
+                else
+                    echo -e "Admin-acks patch finished succ\n"
+                    break
+                fi
+                ((wait_time_loop_var+=1))
+            done
+            [ "${wait_time_loop_var}" -eq 5 ] && echo >&2 "Admin-acks patch not completed in 5 minutes, exiting" && exit 1
+        fi
+    fi
+    set +x
+}
+
+# Upgrade the cluster to target release
+# Parameters: 1- by-digest pull spec, 2- by-tag pull spec
+function upgrade() {
+    set -x
+    local err
+    
+    if ! err="$(./oc adm upgrade --to-image="$1" 2>&1 >/dev/null)" && [[ $err == *"you must pass --allow-explicit-upgrade to continue"* ]]; then
+        echo "Check on oc adm upgrade --to-image missing --allow-explicit-upgrade PASSED"
+    else
+        echo >&2 "Check on oc adm upgrade --to-image missing --allow-explicit-upgrade FAILED: $err, exiting" && exit 1
+    fi
+    local current_image; current_image="$(./oc get clusterversion version -o json|jq -r '.status.history[0].image')"
+    [[ $1 == "${current_image}" ]] && echo >&2 "Upgrade with oc adm upgrade --to-image missing --allow-explicit-upgrade starts but expected NOT start, exiting" && exit 1
+    
+    if [[ "${FORCE_UPDATE}" == "false" ]]; then
+        echo "Upgrading cluster to $1..."
+        if ! ./oc adm upgrade --to-image="$1" --allow-explicit-upgrade; then
+            echo >&2 "Failed to start upgrade, exiting" && exit 1
+        fi
+    else
+        echo "Force upgrading cluster to $2..."
+        if ! ./oc adm upgrade --to-image="$2" --allow-explicit-upgrade --force; then
+            echo >&2 "Failed to start upgrade, exiting" && exit 1
+        fi
+    fi
+
+    set +x
+}
+
+# Monitor the upgrade status
+# Parameter: 1-target release version
+function check_upgrade_status() {
+    local timeout=120
```

Comment:
You can also consider making the `timeout` an environment variable, and set its value depending on job's type